internal/http/middleware.go

```diff
@@ -25,11 +31,96 @@ func LimitBody(next http.Handler) http.Handler {
 	})
 }
 
+type recorder struct {
+	len  int
+	code int
+	http.ResponseWriter
+}
+
+func (r *recorder) Write(b []byte) (int, error) {
+	n, err := r.ResponseWriter.Write(b)
+	r.len += n
+	return n, err
+}
+
+func (r *recorder) WriteHeader(statusCode int) {
+	r.code = statusCode
+	r.ResponseWriter.WriteHeader(statusCode)
+}
+
 // Log logs information about incoming HTTP requests.
-func Log(next http.Handler, l *log.Logger) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		l.Printf("Info: %s request from %s to %s", r.Method, r.RemoteAddr, r.URL.String())
+func Log(next http.Handler, l *logrus.Logger) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		rr := &recorder{ResponseWriter: w}
 
-		next.ServeHTTP(w, r)
-	})
+		start := time.Now()
+		next.ServeHTTP(rr, r)
+		end := time.Now()
+
+		l.WithFields(logrus.Fields{
+			"method":       r.Method,
+			"remote_addr":  r.RemoteAddr,
+			"url":          r.URL.String(),
+			"start_time":   start.Unix(),
+			"request_time": end.Sub(start).Seconds(),
+			"status_code":  rr.code,
+			"body_size":    rr.len,
+			"user_id":      GetSubject(r),
+			"admin":        GetRoles(r).IsAdmin,
+		}).Info("got request")
+	}
+}
+
+// Auth returns a middleware used for jwt authentication.
+func Auth(next http.HandlerFunc, jwtSecret []byte) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("Authentication") == "" {
+			next(w, r)
+			return
+		}
+
+		ss := strings.TrimPrefix(r.Header.Get("Authentication"), "Bearer ")
+		token, err := jwt.ParseWithClaims(ss, &Claims{}, func(token *jwt.Token) (interface{}, error) {
+			if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+				return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+			}
+
+			return jwtSecret, nil
+		})
+		if err != nil {
+			Error(w, http.StatusForbidden)
```

Comment:
I think this should be a 401 rather than a 403 since it's more a "you screwed up your authentication" then "you got authenticated correctly but you're not allowed to access this".